runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java

```diff
@@ -218,19 +220,55 @@ public TxnContext txn(@Nonnull final String namespace, IsolationLevel isolationL
+     * Return the address of the latest update made in this table.
+     *
+     * Note: we can't deliberately return the tail of the stream map
+     * as this entry might be a HOLE, we need to filter and return only that
+     * corresponding to the last DATA entry.
      *
      * @param namespace - namespace that this table belongs to.
      * @param tableName - table name of this table without the namespace prefixed in.
      * @return stream tail of this table.
      */
     public long getHighestSequence(@Nonnull final String namespace,
                                    @Nonnull final String tableName) {
-        return this.runtime.getSequencerView().query(
-                CorfuRuntime.getStreamID(
-                        TableRegistry.getFullyQualifiedTableName(namespace, tableName)
-                )
-        );
+
+        // Overall strategy:
+        // (1) Retrieve streams full address map from sequencer
+        // (2) Read entries in reverse to confirm it's a DATA type entry (for optimization
+        // read in batches). A batch size of 4 is guaranteed to suit one edge case case scenario: i.e.,
+        // consecutive checkpoints enforcing holes (3 rounds of CP accumulate before trim happens == 3 holes)
+
+        long startTime = System.nanoTime();
+        StreamAddressSpace streamAddressSpace = this.runtime.getSequencerView()
+                .getStreamAddressSpace(new StreamAddressRange(CorfuRuntime.getStreamID(
+                        TableRegistry.getFullyQualifiedTableName(namespace, tableName)), Address.MAX, Address.NON_ADDRESS));
+
+        int numBatches = 0;
+
+        if (streamAddressSpace.size() != 0) {
+            NavigableSet<Long> addresses = streamAddressSpace.copyAddressesToSet(Address.MAX).descendingSet();
+            Iterable<List<Long>> batches = Iterables.partition(addresses,
+                    runtime.getParameters().getHighestSequenceNumberBatchSize());
+
+            for (List<Long> batch : batches) {
+                numBatches++;
+                Map<Long, ILogData> entries = runtime.getAddressSpaceView().read(batch);
+                for (Long address : batch) {
+                    if (!entries.get(address).isHole()) {
+                        corfuStoreMetrics.recordBatchReads(numBatches);
+                        corfuStoreMetrics.recordNumberReads(numBatches*runtime.getParameters().getHighestSequenceNumberBatchSize());
+                        corfuStoreMetrics.recordHighestSequenceNumberDuration(System.nanoTime() - startTime);
+                        return address;
+                    }
+                }
+            }
+        }
+
+        corfuStoreMetrics.recordBatchReads(numBatches);
+        corfuStoreMetrics.recordNumberReads(numBatches*runtime.getParameters().getHighestSequenceNumberBatchSize());
+        corfuStoreMetrics.recordHighestSequenceNumberDuration(System.nanoTime() - startTime);
+        return streamAddressSpace.getTrimMark();
```

Comment:
can we add a log line to indicate that we are returning the trim mark? Hopefully this should only happen for totally empty streams.

Reply:
sure, I'm adding this at DEBUG level as I'm guessing this is only for debugging purposes. Otherwise, not sure if for the case of search it could end up being called too many times, cause I'm guessing that on each read they call this API. If you think it should be a different level let me know...